### PR TITLE
Feat/add image threshold

### PR DESCRIPTION
**Pull request by njneeteson**
### Overview

Added a wrapper for the VTK filter vtkImageThreshold for uniform data (vtkImageData)

Resolves: #2157 

#### Filters:

In pyvista/core/filters/uniform_grid.py:

Added a new method 'image_threshold' with a positional argument 'threshold' that can take a single int/float or length 2 sequence of int/float, keyword arguments for what to use to replace scalar values that do or don't meet the threshold criteria, and the standard additional keyword arguments for methods that act on scalar data.

The new method takes a uniform grid and a threshold specification, and it modifies the active/specified scalar data, optionally replacing values that do or don't meet the threshold specified (below/above or between). Another uniform grid dataset is returned.

#### Tests:

In tests/test_filters.py:

1. Test that the output is the right type, with and without specifying the scalars to act on.
2. Check that the right results are produced when thresholding with a single value (below/above) with all 4 combinations of replacing values in/out or not.
3. Same as 2 but for thresholding between two values.

### CODE REVIEW

**Comment by njneeteson** [comment]
Please wait on review / merge until I add two more tests to cover the lines that codecov marked.

File: pyvista/core/filters/data_set.py [hunk 1]
```diff
@@ -431,6 +431,7 @@ def clip_surface(self, surface, invert=True, value=0.0,
                                                     progress_bar=progress_bar)
         return result
 
+
```
Comment (akaszynski) [inline pyvista/core/filters/data_set.py:434]: Only one space between methods.
```suggestion
```

File: pyvista/core/filters/uniform_grid.py [hunk 2]
```diff
@@ -156,3 +158,77 @@ def extract_subset(self, voi, rate=(1, 1, 1), boundary=False, progress_bar=False
...
+        scalars : str, optional
+            Name of scalars to process. Defaults to currently active scalars.
+
+        preference : str, optional
+            When scalars is specified, this is the preferred array
+            type to search for in the dataset.  Must be either
+            ``'point'`` or ``'cell'``.
+
+        progress_bar : bool, optional
+            Display a progress bar to indicate progress. Default False.
```
Comment (akaszynski) [inline pyvista/core/filters/uniform_grid.py:198]:
```suggestion
            Display a progress bar to indicate progress. Default ``False``.
```

File: pyvista/core/filters/uniform_grid.py [hunk 3]
```diff
@@ -156,3 +158,77 @@ def extract_subset(self, voi, rate=(1, 1, 1), boundary=False, progress_bar=False
...
         return fixed
+
+    def image_threshold(self, threshold, in_value=1, out_value=0,
+                        scalars=None, preference='points', progress_bar=False):
+        """Apply a threshold to scalar values in a uniform grid.
+
+        If a single value is given for threshold, scalar values above or equal
+        to the threshold are ``'in'`` and scalar values below the threshold are ``'out.'``
+        If two values are given for threshold (sequence) then values equal to
+        or between the two values are ``'in'`` and values outside the range are ``'out.'``
```
Comment (akaszynski) [inline pyvista/core/filters/uniform_grid.py:169]:
```suggestion
        to the threshold are ``'in'`` and scalar values below the threshold are ``'out'``.
        If two values are given for threshold (sequence) then values equal to
        or between the two values are ``'in'`` and values outside the range are ``'out'``.
```

File: tests/test_filters.py [hunk 4]
```diff
@@ -1451,6 +1451,84 @@ def test_median_smooth_outlier():
                           volume_outlier_smoothed.point_data['point_data'])
 
 
+def test_image_threshold_output_type():
+    threshold = 10 # 'random' value
```
Comment (akaszynski) [inline tests/test_filters.py:1455]:
```suggestion
    threshold = 10  # 'random' value
```

File: tests/test_filters.py [hunk 5]
```diff
@@ -1451,6 +1451,84 @@ def test_median_smooth_outlier():
...
+    volume = examples.load_uniform()
+    volume_thresholded = volume.image_threshold(threshold)
+    assert isinstance(volume_thresholded, pyvista.UniformGrid)
+    volume_thresholded = volume.image_threshold(threshold, scalars='Spatial Point Data')
+    assert isinstance(volume_thresholded, pyvista.UniformGrid)
+
+
+def test_image_threshold_wrong_threshold_length():
+    threshold = (10,10,10) # tuple with too many values
```
Comment (akaszynski) [inline tests/test_filters.py:1464]:
```suggestion
    threshold = (10, 10, 10)  # tuple with too many values
```

File: tests/test_filters.py [hunk 6]
```diff
@@ -1451,6 +1451,84 @@ def test_median_smooth_outlier():
...
+    volume = examples.load_uniform()
+    with pytest.raises(ValueError):
+        volume_thresholded = volume.image_threshold(threshold)
+
+
+def test_image_threshold_wrong_threshold_type():
+    threshold = {'min':0, 'max':10} # dict thresh
```
Comment (akaszynski) [inline tests/test_filters.py:1471]:
```suggestion
    threshold = {'min': 0, 'max': 10}  # dict thresh
```

File: tests/test_filters.py [hunk 7]
```diff
@@ -1451,6 +1451,84 @@ def test_median_smooth_outlier():
...
+    with pytest.raises(TypeError):
+        volume_thresholded = volume.image_threshold(threshold)
+
+
+@pytest.mark.parametrize('in_value', [1, None])
+@pytest.mark.parametrize('out_value', [0, None])
+def test_image_threshold_upper(in_value,out_value):
+    threshold = 0 # 'random' value
+    array_shape = (3,3,3)
+    in_value_location = (1,1,1)
```
Comment (akaszynski) [inline tests/test_filters.py:1482]:
```suggestion
    threshold = 0  # 'random' value
    array_shape = (3, 3, 3)
    in_value_location = (1, 1, 1)
```

File: tests/test_filters.py [hunk 8]
```diff
@@ -1451,6 +1451,84 @@ def test_median_smooth_outlier():
...
+    point_data = np.ones(array_shape)
+    in_value_mask = np.zeros(array_shape,dtype=bool)
+    in_value_mask[in_value_location] = True
+    point_data[in_value_mask] = 100 # the only 'in' value
+    point_data[~in_value_mask] = -100 # out values
```
Comment (akaszynski) [inline tests/test_filters.py:1487]:
```suggestion
    point_data[in_value_mask] = 100  # the only 'in' value
    point_data[~in_value_mask] = -100  # out values
```

File: tests/test_filters.py [hunk 9]
```diff
@@ -1451,6 +1451,84 @@ def test_median_smooth_outlier():
...
+    volume.point_data['point_data'] = point_data.flatten(order='F')
+    point_data_thresholded = point_data.copy()
+    if in_value is not None:
+        point_data_thresholded[in_value_mask] = in_value
+    if out_value is not None:
+        point_data_thresholded[~in_value_mask] = out_value
+    volume_thresholded = volume.image_threshold(threshold,
+                                                in_value = in_value,
+                                                out_value = out_value
+                                                )
```
Comment (akaszynski) [inline tests/test_filters.py:1498]:
```suggestion
    volume_thresholded = volume.image_threshold(
        threshold,
        in_value=in_value,
        out_value=out_value
    )
```

File: tests/test_filters.py [hunk 10]
```diff
@@ -1451,6 +1451,84 @@ def test_median_smooth_outlier():
...
```
Comment (akaszynski) [inline tests/test_filters.py:1527]:
```suggestion
    volume_thresholded = volume.image_threshold(
        threshold,
        in_value=in_value,
        out_value=out_value
    )
```

File: tests/test_filters.py [hunk 11]
```diff
@@ -1451,6 +1451,84 @@ def test_median_smooth_outlier():
...
+    threshold = [-10, 10] # 'random' values
+    array_shape = (3,3,3)
+    in_value_location = (1,1,1)
+    low_value_location = (1,1,2)
+    point_data = np.ones(array_shape)
+    in_value_mask = np.zeros(array_shape,dtype=bool)
+    in_value_mask[in_value_location] = True
+    point_data[in_value_mask] = 0 # the only 'in' value
+    point_data[~in_value_mask] = 100 # out values
+    point_data[low_value_location] = -100 # add a value below the threshold also
```
Comment (akaszynski) [inline tests/test_filters.py:1515]:
```suggestion
    point_data[in_value_mask] = 0  # the only 'in' value
    point_data[~in_value_mask] = 100  # out values
    point_data[low_value_location] = -100  # add a value below the threshold also
```

File: tests/test_filters.py [hunk 12]
```diff
@@ -1451,6 +1451,84 @@ def test_median_smooth_outlier():
...
+                          point_data_thresholded.flatten(order='F'))
+
+
+@pytest.mark.parametrize('in_value', [1, None])
+@pytest.mark.parametrize('out_value', [0, None])
+def test_image_threshold_between(in_value,out_value):
+    threshold = [-10, 10] # 'random' values
+    array_shape = (3,3,3)
+    in_value_location = (1,1,1)
+    low_value_location = (1,1,2)
```
Comment (akaszynski) [inline tests/test_filters.py:1509]:
```suggestion
    array_shape = (3, 3, 3)
    in_value_location = (1, 1, 1)
    low_value_location = (1, 1, 2)
```

File: tests/test_filters.py [hunk 13]
```diff
@@ -1451,6 +1451,86 @@ def test_median_smooth_outlier():
...
+    array_shape = (3, 3, 3)
+    in_value_location = (1, 1, 1)
+    low_value_location = (1, 1, 2)
+    point_data = np.ones(array_shape)
+    in_value_mask = np.zeros(array_shape,dtype=bool)
+    in_value_mask[in_value_location] = True
+    point_data[in_value_mask] = 0  # the only 'in' value
+    point_data[~in_value_mask] = 100  # out values
+    point_data[low_value_location] = -100  # add a value below the threshold also
+    point_data
```
Comment (akaszynski) [inline tests/test_filters.py:1517]:
```suggestion

```

File: pyvista/core/filters/uniform_grid.py [hunk 14]
```diff
@@ -156,3 +158,77 @@ def extract_subset(self, voi, rate=(1, 1, 1), boundary=False, progress_bar=False
...
+            ``'point'`` or ``'cell'``.
+
+        progress_bar : bool, optional
+            Display a progress bar to indicate progress. Default ``False``.
+
+        Returns
+        -------
+        pyvista.UniformGrid
+            Dataset with the specified scalars thresholded.
+        """
```
Comment (akaszynski) [inline pyvista/core/filters/uniform_grid.py:204]:
```suggestion

        Examples
        --------
        Demonstrate image threshold on an example dataset. First, plot
        the example dataset with the active scalars.

        >>> from pyvista import examples
        >>> uni = examples.load_uniform()
        >>> uni.plot()

        Now, plot the image threshold with ``threshold=100``. Note how
        values above the threshold are 1 and below are 0.

        >>> ithresh = uni.image_threshold(100)
        >>> ithresh.plot()

        """
```

**Review by akaszynski — APPROVED** [review]
I've made a few whitespace change and added an example, but overall I think it's great.

Thanks for your work!

**Pull request merged**